src/libxtp/gwbse/gw.cc

```diff
@@ -203,83 +194,146 @@ void GW::CalculateGWPerturbation() {
       }
     }
   }
-
+  _Sigma_c.diagonal() = _sigma->CalcCorrelationDiag(frequencies);
   PrintGWA_Energies();
 }
 
-Eigen::VectorXd GW::SolveQP_Grid(const Eigen::VectorXd& frequencies) const {
-  const Index qptotal = _opt.qpmax - _opt.qpmin + 1;
-  const double range =
-      _opt.qp_grid_spacing * (double)(_opt.qp_grid_steps - 1) / 2.0;
-  const Eigen::VectorXd intercept =
+Eigen::VectorXd GW::getGWAResults() const {
+  return _Sigma_x.diagonal() + _Sigma_c.diagonal() - _vxc.diagonal() +
+         _dft_energies.segment(_opt.qpmin, _qptotal);
+}
+
+Eigen::VectorXd GW::SolveQP(const Eigen::VectorXd& frequencies) const {
+  const Eigen::VectorXd intercepts =
       _dft_energies.segment(_opt.qpmin, _qptotal) + _Sigma_x.diagonal() -
       _vxc.diagonal();
   Eigen::VectorXd frequencies_new = frequencies;
+  Eigen::ArrayXi converged = Eigen::ArrayXi::Zero(_qptotal);
```

Comment:
Why not a vector of booleans? Eigen has nice reduction methods like .any() and .all(), which you could use later to check whether all states have converged.

Reply:
Good point.